src/contracts/middleware/StakeRegistry.sol

```diff
@@ -0,0 +1,481 @@
+            }
+        }
+
+        // record stake updates in the EigenLayer Slasher
+        for (uint i = 0; i < operators.length;) {
+            serviceManager.recordStakeUpdate(operators[i], uint32(block.number), serviceManager.latestServeUntilBlock(), prevElements[i]);
+            unchecked {
+                ++i;
+            }
+        }     
+    }
+
+    // INTERNAL FUNCTIONS
+
+    function _registerOperator(address operator, bytes32 operatorId, bytes memory quorumNumbers) internal {
+        uint8 quorumNumbersLength = uint8(quorumNumbers.length);
+        // check the operator is registering for only valid quorums
+        require(uint8(quorumNumbers[quorumNumbersLength - 1]) < quorumCount, "StakeRegistry._registerStake: greatest quorumNumber must be less than quorumCount");
+        OperatorStakeUpdate memory _newTotalStakeUpdate;
+        // add the `updateBlockNumber` info
+        _newTotalStakeUpdate.updateBlockNumber = uint32(block.number);
+        // for each quorum, evaluate stake and add to total stake
+        for (uint8 quorumNumbersIndex = 0; quorumNumbersIndex < quorumNumbersLength;) {
+            // get the next quourumNumber
+            uint8 quorumNumber = uint8(quorumNumbers[quorumNumbersIndex]);
+            // evaluate the stake for the operator
+            // since we don't use the first output, this will use 1 extra sload when deregistered operator's register again
+            (, uint96 stake) = _updateOperatorStake(operator, operatorId, quorumNumber);
+            // @JEFF: This reverts pretty late, but i think that's fine. wdyt?
+            // check if minimum requirement has been met, will be 0 if not
+            require(stake != 0, "StakeRegistry._registerStake: Operator does not meet minimum stake requirement for quorum");
+            // add operator stakes to total stake before update (in memory)
+            uint256 _totalStakeHistoryLength = _totalStakeHistory[quorumNumber].length;
+            if (_totalStakeHistoryLength != 0) {
+                // only add the stake if there is a previous total stake
+                // overwrite `stake` variable
+                stake += _totalStakeHistory[quorumNumber][_totalStakeHistoryLength - 1].stake;
+            }
+            _newTotalStakeUpdate.stake = stake;
+            // update storage of total stake
+            _recordTotalStakeUpdate(quorumNumber, _newTotalStakeUpdate);
+            unchecked {
+                ++quorumNumbersIndex;
+            }
+        }
+    }
+
+    /**
+     * @notice Removes the stakes of the operator
+     */
+    function _deregisterOperator(bytes32 operatorId, bytes memory quorumNumbers) internal {
+        uint8 quorumNumbersLength = uint8(quorumNumbers.length);
+        // check the operator is deregistering from only valid quorums
+        require(uint8(quorumNumbers[quorumNumbersLength - 1]) < quorumCount, "StakeRegistry._registerStake: greatest quorumNumber must be less than quorumCount");
```

Comment:
is it possible to remove quorums? seems like this _could_ lead to a bug where it becomes impossible to deregister (at least, if it's possible to remove quorums).

Reply:
it is not

Reply:
If the flow to get here is permissioned, perhaps just make it a precondition that we've already checked that the provided `quorumNumbers` match the operator's current bitmap exactly?

Reply:
the quoted bit is part of _de_registration!

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/76/commits/f97ac404f08237474a68639a1037f92243dffc0f